design/ozone.md

```diff
@@ -0,0 +1,246 @@
+
+### Kubernetes resource generation
+
+Ozone provides examples K8s resources files and transformation definitions using [flekszible](https://github.com/elek/flekszible). This is a composition based k8s resource generator where the transformations can be reused.
+
+Let's say we have a simple kubernetes resource:
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: datanode
+  labels:
+    app.kubernetes.io/component: ozone
+spec:
+  selector:
+    matchLabels:
+      app: ozone
+      component: datanode
+  template:
+    metadata:
+      labels:
+        app: ozone
+        component: datanode
+    spec:
+      containers:
+      - name: datanode
+        image: "apache/hadoop:0.5.0-SNAPSHOT"
+        args: ["ozone","datanode"]
+        ports:
+           - containerPort: 9870
+             name: rpc
+```
+
+Prometheus annotations can be added by a simple rule:
+
+```yaml
+type: Add
+path: ["spec","template","spec","metadata","annotations"]
+trigger:
+   metadata:
+      name: datanode
+value:
+   prometheus.io/scrape: "true"
+   prometheus.io/port: "9882"
+   prometheus.io/path: "/prom"
+```
+
+This transformation can be applied to any of the k8s resources.
+
+Until now it is very similar to Kustomize, but here we have more predefined transformations (DaemonSet to StatefulSet converter, NodePort service generator, value replacer).
+
+Even better the transformations can be reused. The previous definition can be reused by putting it to the right directory:
+
+```yaml
+name: ozone/prometheus
+description: Enable prometheus monitoring in Ozone
+---
+- type: Add
+  path: ["spec","template","spec","metadata","annotations"]
+  trigger:
+     metadata:
+        name: datanode
+  value:
+     prometheus.io/scrape: "true"
+     prometheus.io/port: "9882"
+     prometheus.io/path: "/prom"
+```
+
+After this step, you can use it as
+
+```yaml
+type: ozone/prometheus
+```
+
+Or from golang API: 
+
+```golang
+
+context, err := Initialize("../testdata/api")
+if err != nil {
+   return err
+}
+
+err = context.AppendCustomProcessor("ozone/prometheus", make(map[string]string))
+if err != nil {
+   return err
+}
+
+context.Render()
+
+s, err := context.Resources()[0].Content.ToString()
+if err != nil {
+   return err
+}
+
+assert.Contains(t, s, "prometheus.io/scrape: \"true\"")
+```
+
+### Create k8s resource from operator
+
+One easy way to integrate the exising resource generation of Ozone is to use it from golang API. This is similar to the approach of [ceph csi](https://github.com/rook/rook/tree/master/pkg/operator/ceph/csi) which reads the k8s template files from the disk.
+
+The main diference here is that before the final adjustment we can reuse the basic transformations which are already implemented by ozone.
```

Comment:
What's another example of the transformation? Will the transformation change attributes in the template such as the namespace, the labels, or settings to apply? If not very many attributes need to change it should work fine. If many attributes in the resources need to be changed, you might consider using the k8s api directly, which is used by most storage providers to give full flexibility. 

The CSI drivers just use the templates since they don't need to change much. When these were added I was thinking it was a temporary situation until we converted them to the k8s api, but they still seem valid since they are picked up from another repo (ceph-csi) and they are mostly static.

Reply:
It depends from the transformation. But some of the transformations are defined by the Ozone project itself (for example to turn on prometheus integration or distributued tracing) 

I think it can be a good rule of thumb:

 * Reuse existing transformations which are provided by the Ozone project.
 * Modify the results only with using k8s api  _from_ the rook operator.